### PR TITLE
Implement Pipedrive OAuth provider

### PR DESCRIPTION
**Pull request by denis-goncharenko**
This feature allows users to authenticate using their Pipedrive credentials.

### CODE REVIEW

File: src/AspNet.Security.OAuth.Pipedrive/AspNet.Security.OAuth.Pipedrive.csproj [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>$(DefaultNetCoreTargetFramework)</TargetFrameworks>
+  </PropertyGroup>
+
+  <!-- TODO Enable once this provider is published to NuGet.org -->
+  <PropertyGroup>
+    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Pipedrive/AspNet.Security.OAuth.Pipedrive.csproj:9]:
```suggestion
    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
    <PackageValidationBaselineVersion>8.0.1</PackageValidationBaselineVersion>
```

File: src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+        using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving the user profile.");
+        }
+
+        using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationHandler.cs:42]:
```suggestion
        using var stream = await response.Content.ReadAsStreamAsync(Context.RequestAborted);
        using var payload = await JsonDocument.ParseAsync(stream, cancellationToken: Context.RequestAborted);
```

File: src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        CallbackPath = PipedriveAuthenticationDefaults.CallbackPath;
+
+        AuthorizationEndpoint = PipedriveAuthenticationDefaults.AuthorizationEndpoint;
+        TokenEndpoint = PipedriveAuthenticationDefaults.TokenEndpoint;
+        UserInformationEndpoint = PipedriveAuthenticationDefaults.UserInformationEndpoint;
+
+        Scope.Add("base");
+
+        ClaimActions.MapCustomJson(ClaimTypes.NameIdentifier,
+            user => user.GetProperty("data").GetString("id"));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationOptions.cs]:
```suggestion
        ClaimActions.MapCustomJson(ClaimTypes.NameIdentifier, user => user.GetProperty("data").GetString("id"));
```

**Review by martincostello — COMMENTED** [review]
Just a few minor comments - I assume you've tested this provider works against the production Pipedrive service?

**Comment by denis-goncharenko** [comment]
> Just a few minor comments - I assume you've tested this provider works against the production Pipedrive service?

I've tested the Pipedrive provider against the production service and can confirm that it works fine.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for the contribution - the changes look good. I'll look to merge this later today (or tomorrow) when I have some time to put aside to manage the release.

File: src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving the user profile.");
+        }
+
+        await using var stream = await response.Content.ReadAsStreamAsync(Context.RequestAborted);
+        using var payload = await JsonDocument.ParseAsync(stream, cancellationToken: Context.RequestAborted);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationHandler.cs]: Sorry - we've reverted this change. Can you do the same in your other PRs too please?

```suggestion
        using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
```

Comment (denis-goncharenko) [reply src/AspNet.Security.OAuth.Pipedrive/PipedriveAuthenticationHandler.cs]: Done for all PRs.

**Review by kevinchalet — APPROVED** [review]
Thanks for your PR!

**Pull request merged**